src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,259 @@
+      this.v3PoolProvider!
+    );
+
+    const gasCostUSD = await getGasCostInUSD(nativeCurrency, usdPool, costNativeCurrency)
+    
+    let gasCostQuoteToken = costNativeCurrency
+
+    // get fee in terms of quote token
+    if (!(quoteToken.wrapped.equals(nativeCurrency))) {
+      const nativePool = await getHighestLiquidityV3NativePool(
+        quoteToken.wrapped,
+        this.v3PoolProvider!
+      );
+      if(!nativePool) {
+        log.info('Could not find a pool to convert the cost into the quote token')
+        gasCostQuoteToken = CurrencyAmount.fromRawAmount(quoteToken.wrapped, 0);
+      } else {
+        gasCostQuoteToken = await getGasCostInQuoteToken(quoteToken.wrapped, nativePool, costNativeCurrency)
+      }
+    }
+
+    // Adjust quote for gas fees
+    let quoteGasAdjusted: CurrencyAmount
+    if(tokenIn.wrapped.equals(quoteToken.wrapped)) {
+      // Exact output - need more of tokenIn to get the desired amount of tokenOut
+      quoteGasAdjusted = route.quote.add(gasCostQuoteToken)
+    } else {
+      // Exact input - can get less of tokenOut due to fees
+      quoteGasAdjusted = route.quote.subtract(gasCostQuoteToken)
+    }
+
+    return {...route, estimatedGasUsedUSD: gasCostUSD, estimatedGasUsedQuoteToken: gasCostQuoteToken, quoteGasAdjusted:quoteGasAdjusted}
+  }
+}
+export class TenderlySimulator {
+  private tenderlyBaseUrl: string
+  private tenderlyUser: string
+  private tenderlyProject: string
+  private tenderlyAccessKey: string
+
+  constructor(tenderlyBaseUrl: string, tenderlyUser: string, tenderlyProject: string, tenderlyAccessKey: string) {
+    this.tenderlyBaseUrl = tenderlyBaseUrl
+    this.tenderlyUser = tenderlyUser
+    this.tenderlyProject = tenderlyProject
+    this.tenderlyAccessKey = tenderlyAccessKey
+  }
+
+  public async simulateTransaction(
+    fromAddress: string,
+    route: SwapRoute,
+  ):Promise<BigNumber> {
+    const tokenIn = route.trade.inputAmount.currency.wrapped
+    if([ChainId.CELO, ChainId.CELO_ALFAJORES].includes(tokenIn.chainId)) {
+      const msg = "Celo not supported by Tenderly!"
+      log.info(msg)
+      throw new Error(msg)
+    }
+
+    if(!route.methodParameters) {
+      throw new Error("No calldata provided to simulate transaction")
+    }
+    const { calldata } = route.methodParameters
+    log.info(
+      {
+        calldata: route.methodParameters.calldata,
+        fromAddress: fromAddress,
+        chainId: tokenIn.chainId,
+        tokenInAddress: tokenIn.address,
+      },
+      'Simulating transaction via Tenderly'
+    )
+
+    const approve = {
+      network_id: tokenIn.chainId,
+      input: APPROVE_TOKEN_FOR_TRANSFER,
+      to: tokenIn.address,
+      value: "0",
+      from: fromAddress,
+      gasPrice: "0",
+      gas: 30000000,
+    }
+
+    const swap = {
+      network_id: tokenIn.chainId,
+      input: calldata,
+      to: V3_ROUTER2_ADDRESS,
+      value: BigNumber.from(route.methodParameters.value).toString(),
```

Comment:
Pretty sure simulation will fail if the fromAddress does not have this much native eth? 

I think the integ test accounts have huge balances of ETH so the tests are probably passing incorrectly?

Reply:
We should only set this if the original token in specified (i.e. the Currency object passed into the top level route() call) is native